src/composite.c

```diff
@@ -72,40 +82,22 @@ int Composite(VipsObject *context, VipsImage *src, VipsImage *dst, VipsImage **o
   //
   // Wikipedia:
   // out_rgb = (src_rgb * src_a + dst_rgb * dst_a * (1 - src_a)) / out_a
+  //                                                ^^^^^^^^^^^
+  //                                                    t0
   //
-  // `vips_ifthenelse` with `blend=TRUE`: http://bit.ly/1KoSsga
-  // out = (cond / 255) * in1 + (1 - cond / 255) * in2
-  //
-  // Substitutions:
+  // Omit division by `out_a` since `Compose` is supposed to output a
+  // premultiplied RGBA image as reversal of premultiplication is handled
+  // externally.
   //
-  //     cond --> src_a
-  //     in1 --> src_rgb
-  //     in2 --> dst_rgb * dst_a (premultiplied destination RGB)
-  //
-  // Finally, manually divide by `out_a` to unpremultiply the RGB channels.
-  // Failing to do so results in darker than expected output with low
-  // opacity images.
-  //
-  VipsImage *dstRGBPremultiplied;
-  if (vips_multiply(dstRGB, dstAlphaNormalized, &dstRGBPremultiplied, NULL))
-    return -1;
-
-  vips_object_local(context, dstRGBPremultiplied);
-
+  VipsImage *t2;
   VipsImage *outRGBPremultiplied;
-  if (vips_ifthenelse(srcAlpha, srcRGB, dstRGBPremultiplied,
-      &outRGBPremultiplied, "blend", TRUE, NULL))
+  if (vips_multiply(dstRGBPremultiplied, t0, &t2, NULL) ||
+      vips_add(srcRGBPremultiplied, t2, &outRGBPremultiplied, NULL))
     return -1;
 
+  vips_object_local(context, t2);
   vips_object_local(context, outRGBPremultiplied);
```

Comment:
libvips uses this style, as a possible alternative:

```C
VipsImage **t = vips_object_local_array(..);
VipsImage *in;

in = ..->input;

if( vips_thing(in, &t[0], NULL ) || 
    vips_thing2(t[0], &t[1], NULL )
    return -1;
in = t[2];

if( vips_thing3(in, &t[2], NULL ) || 
    vips_thing4(t[2], &t[3], NULL )
    return -1;
in = t[3];

if (vips_image_write(in, out))
    return -1;

return 0;
```

This has some nice properties:

* you can use `.. || ..` to chain operations together, so it's concise
* guaranteed leak-free
* you only need to keep track of the annoying `t[]` variables with a `.. || ..` block -- the blocks are independent and joined with properly named variables
* because blocks are independent, you can reorder them safely: you could swap these two over and it would all work